File: ios/WhoCalled/Services/WhoCalledAPI.swift
import Foundation

// MARK: - API models

struct ListNumberDTO: Codable {
  let phone: String
  let status: String
  let spamScore: Int
  let category: String
  let source: String?
  let updatedAt: String
}

struct ListPatternDTO: Codable {
  let pattern: String
  let status: String
  let category: String
  let source: String
  let name: String?
  let updatedAt: String
}

struct ListResponseDTO: Codable {
  let serverTime: String
  let count: Int
  let next: String?
  let numbers: [ListNumberDTO]
  let patterns: [ListPatternDTO]?
  // Phones to drop from the local cache (healed / deleted upstream).
  let deleted: [String]?
}

/// GET /lists/access — short-lived token gating the list download.
struct ListAccessDTO: Codable {
  let token: String
  let expiresIn: Int
}

struct LookupFrequencyDTO: Codable {
  let last24h: Int
  let last7d: Int
  let last30d: Int
  let last1y: Int
}

/// Dominant community reason ("why did it call?") for a number.
struct TopReasonDTO: Codable {
  let category: String  // telemarketing | scam | robocall | unknown
  let count: Int
  let share: Int  // 0..100 — share of spam reports
}

struct LookupResponseDTO: Codable {
  let phone: String
  let spamScore: Int
  let status: String
  let category: String?
  let source: String?
  let reportCountSpam: Int?
  let reportCountLegit: Int?
  let frequency: LookupFrequencyDTO?
  let topReason: TopReasonDTO?
  let firstReportedAt: String?
  let lastReportedAt: String?
}

/// One number in the "qui montent" trending list.
struct TrendingNumberDTO: Codable, Identifiable {
  let phone: String
  let reportCount: Int
  let last24h: Int
  let velocity: Int
  let spamScore: Int
  let status: String
  let category: String
  let source: String
  let topReason: TopReasonDTO?

  var id: String { phone }
}

struct TrendingResponseDTO: Codable {
  let serverTime: String?
  let window: Int?
  let count: Int?
  let numbers: [TrendingNumberDTO]
}

struct ReportRequestDTO: Codable {
  let phone: String
  let deviceId: String
  let vote: String  // "spam" | "legit"
  let category: String?
  let locale: String
}

/// GET /stats — public reassurance figures shown on the Home banner.
struct StatsResponseDTO: Codable {
  let serverTime: String?
  let coveredNumbers: Int?  // big "numéros couverts" headline
  let communityCount: Int?  // honest count of crowdsourced numbers
  let arcepPatternCount: Int?  // number of ARCEP block ranges
  let lastUpdate: String?  // ISO-8601 of freshest data, or null
  let country: String?  // echoed dial code, or null if global
  let countryNumbers: Int?  // numbers for that country
  let countryArcepPatternCount: Int?  // ARCEP ranges for that country
}

/// POST /game/score — per-game leaderboard (score-based, higher = better).
struct GameScoreRequestDTO: Codable {
  let deviceId: String
  let score: Int
  let waves: Int
  let day: Int
  let game: String  // defense | trace
}

/// One TRACE grid served by the backend (region ids + the unique solution).
struct TraceGrid: Codable {
  let n: Int
  let region: [[Int]]
  let solution: [Int]
}

/// GET /game/puzzle — the day's TRACE sprint (5 grids) or a training grid.
struct TracePuzzleSetDTO: Codable {
  let day: Int?
  let game: String?
  let mode: String?
  let grids: [TraceGrid]
}

struct GameScoreResponseDTO: Codable {
  let players: Int
  let bestScore: Int
  let medianScore: Int
  let yourScore: Int
  let rank: Int
  let topPercent: Int
}

struct LeaderboardEntryDTO: Codable, Identifiable { let rank: Int; let score: Int; let waves: Int; var id: Int { rank } }
struct LeaderboardYouDTO: Codable { let score: Int; let waves: Int; let rank: Int; let topPercent: Int }
struct LeaderboardResponseDTO: Codable {
  let day: Int; let players: Int; let bestScore: Int; let medianScore: Int
  let top: [LeaderboardEntryDTO]; let you: LeaderboardYouDTO?
}
struct HistoryEntryDTO: Codable, Identifiable {
  let day: Int; let score: Int; let waves: Int; let rank: Int; let players: Int; let topPercent: Int
  var id: Int { day }
}
struct HistoryResponseDTO: Codable { let entries: [HistoryEntryDTO] }

struct CommunityResponseDTO: Codable { let total: Int; let week: Int; let goal: Int }

enum APIError: Error {
  case invalidURL
  case http(Int)
  case decoding
}

/// Minimal async HTTP client for the who-called API (URLSession + Codable).
struct WhoCalledAPI {
  /// Short-lived access token gating GET /lists (the download URL is
  /// deliberately temporary — dead a few minutes later if replayed).
  func listsAccess() async throws -> String {
    guard let url = URL(string: "\(AppConstants.apiListsURL)/access") else {
      throw APIError.invalidURL
    }
    let dto: ListAccessDTO = try await get(url)
    return dto.token
  }

  func fetchList(since: String?, country: String? = nil, limit: Int = 5000, token: String) async throws -> ListResponseDTO {
    var components = URLComponents(string: AppConstants.apiListsURL)
    var items = [URLQueryItem(name: "limit", value: String(limit))]
    if let since { items.append(URLQueryItem(name: "since", value: since)) }
    if let country, !country.isEmpty { items.append(URLQueryItem(name: "country", value: country)) }
    items.append(URLQueryItem(name: "token", value: token))
    components?.queryItems = items
    guard let url = components?.url else { throw APIError.invalidURL }
    return try await get(url)
  }

  func lookup(_ phone: String) async throws -> LookupResponseDTO {
    guard let url = URL(string: "\(AppConstants.apiLookupURL)/\(phone)") else {
      throw APIError.invalidURL
    }
    return try await get(url)
  }

  /// "Numéros qui montent" — most-reported numbers ramping up recently.
  func trending(windowDays: Int = 7, limit: Int = 25) async throws -> TrendingResponseDTO {
    var components = URLComponents(string: AppConstants.apiTrendingURL)
    components?.queryItems = [
      URLQueryItem(name: "window", value: String(windowDays)),
      URLQueryItem(name: "limit", value: String(limit)),
    ]
    guard let url = components?.url else { throw APIError.invalidURL }
    return try await get(url)
  }

  /// Shared community-shield figures (total + this week + soft weekly goal).
  func community() async throws -> CommunityResponseDTO {
    guard let url = URL(string: AppConstants.apiStatsURL + "/community") else { throw APIError.invalidURL }
    return try await get(url)
  }

  /// Public reassurance stats. `country` (dial code) adds that country's counts.
  func stats(country: String? = nil) async throws -> StatsResponseDTO {
    var components = URLComponents(string: AppConstants.apiStatsURL)
    if let country, !country.isEmpty {
      components?.queryItems = [URLQueryItem(name: "country", value: country)]
    }
    guard let url = components?.url else { throw APIError.invalidURL }
    return try await get(url)
  }

  /// Connectivity probe for the Settings "mode test" (double-tap version).
  /// Hits the root /health endpoint and returns the round-trip latency in ms.
  func ping() async throws -> Int {
    guard let url = URL(string: AppConstants.apiHealthURL) else { throw APIError.invalidURL }
    var request = URLRequest(url: url)
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    let start = Date()
    let (_, response) = try await URLSession.shared.data(for: request)
    try validate(response)
    return Int(Date().timeIntervalSince(start) * 1000)
  }

  func report(_ body: ReportRequestDTO) async throws {
    guard let url = URL(string: AppConstants.apiReportsURL) else { throw APIError.invalidURL }
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.httpBody = try JSONEncoder().encode(body)
    let (_, response) = try await URLSession.shared.data(for: request)
    try validate(response)
  }

  /// Submit a run for `game` → returns today's anonymous leaderboard position.
  func submitGameScore(score: Int, waves: Int, day: Int, game: String = "defense") async throws -> GameScoreResponseDTO {
    guard let url = URL(string: AppConstants.apiGameScoreURL) else { throw APIError.invalidURL }
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    request.httpBody = try JSONEncoder().encode(
      GameScoreRequestDTO(deviceId: SharedStore.deviceId(), score: score, waves: waves, day: day, game: game))
    let (data, response) = try await URLSession.shared.data(for: request)
    try validate(response)
    do { return try JSONDecoder().decode(GameScoreResponseDTO.self, from: data) }
    catch { throw APIError.decoding }
  }

  /// Anonymous leaderboard for a game + period (day|week|all) + your position.
  func gameLeaderboard(period: String, day: Int, game: String = "defense") async throws -> LeaderboardResponseDTO {
    var c = URLComponents(string: AppConstants.apiGameScoreURL.replacingOccurrences(of: "/score", with: "/leaderboard"))
    c?.queryItems = [
      URLQueryItem(name: "period", value: period),
      URLQueryItem(name: "day", value: String(day)),
      URLQueryItem(name: "game", value: game),
      URLQueryItem(name: "deviceId", value: SharedStore.deviceId()),
    ]
    guard let url = c?.url else { throw APIError.invalidURL }
    return try await get(url)
  }

  /// This device's recent daily results for a game, with per-day rank.
  func gameHistory(days: Int = 14, game: String = "defense") async throws -> HistoryResponseDTO {
    var c = URLComponents(string: AppConstants.apiGameScoreURL.replacingOccurrences(of: "/score", with: "/history"))
    c?.queryItems = [
      URLQueryItem(name: "deviceId", value: SharedStore.deviceId()),
      URLQueryItem(name: "days", value: String(days)),
      URLQueryItem(name: "game", value: game),
    ]
    guard let url = c?.url else { throw APIError.invalidURL }
    return try await get(url)
  }

  /// TRACE: the day's 5-grid sprint, identical for every player.
  func tracePuzzle(day: Int) async throws -> TracePuzzleSetDTO {
    var c = URLComponents(string: AppConstants.apiGamePuzzleURL)
    c?.queryItems = [URLQueryItem(name: "game", value: "trace"), URLQueryItem(name: "day", value: String(day))]
    guard let url = c?.url else { throw APIError.invalidURL }
    return try await get(url)
  }

  /// TRACE: a single unranked training grid of the requested size.
  func traceTraining(size: Int, seed: Int) async throws -> TracePuzzleSetDTO {
    var c = URLComponents(string: AppConstants.apiGamePuzzleURL)
    c?.queryItems = [
      URLQueryItem(name: "game", value: "trace"),
      URLQueryItem(name: "mode", value: "training"),
      URLQueryItem(name: "size", value: String(size)),
      URLQueryItem(name: "seed", value: String(seed)),
    ]
    guard let url = c?.url else { throw APIError.invalidURL }
    return try await get(url)
  }

  /// RGPD: erase all reports made from this device.
  func eraseDevice(_ deviceId: String) async throws {
    let encoded = deviceId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? deviceId
    guard let url = URL(string: "\(AppConstants.apiBaseURL)/privacy/device/\(encoded)") else {
      throw APIError.invalidURL
    }
    var request = URLRequest(url: url)
    request.httpMethod = "DELETE"
    let (_, response) = try await URLSession.shared.data(for: request)
    try validate(response)
  }

  private func get<T: Decodable>(_ url: URL) async throws -> T {
    var request = URLRequest(url: url)
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    let (data, response) = try await URLSession.shared.data(for: request)
    try validate(response)
    do {
      return try JSONDecoder().decode(T.self, from: data)
    } catch {
      throw APIError.decoding
    }
  }

  private func validate(_ response: URLResponse) throws {
    guard let http = response as? HTTPURLResponse else { return }
    guard (200..<300).contains(http.statusCode) else { throw APIError.http(http.statusCode) }
  }
}

File: android/app/src/main/java/com/whocalled/android/network/WhoCalledApi.kt
package com.whocalled.android.network

import com.google.gson.Gson
import com.whocalled.android.BuildConfig
import com.whocalled.android.game.TracePuzzleSet
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * Minimal HTTP client for the who-called API using HttpURLConnection + Gson.
 * No third-party HTTP dependency — keeps the app light (like Saracroche's approach).
 */
class WhoCalledApi(
    private val baseUrl: String = BuildConfig.API_BASE_URL,
    private val gson: Gson = Gson(),
) {

    /**
     * Connectivity probe used by the Settings "mode test" (double-tap version).
     * Hits the root /health endpoint (note: it lives at the host root, NOT under
     * /api/v1) and returns the round-trip latency in ms. Throws on any failure.
     */
    fun ping(): Long {
        // baseUrl is ".../api/v1" → strip the API prefix to reach the host /health.
        val healthUrl = baseUrl.replace(Regex("/api/v\\d+/?$"), "") + "/health"
        val start = System.currentTimeMillis()
        request(URL(healthUrl), "GET", null, Map::class.java)
        return System.currentTimeMillis() - start
    }

    /** Shared community-shield figures (total + this week + soft weekly goal). */
    fun community(): CommunityResponse {
        val url = URL("$baseUrl/stats/community")
        return request(url, "GET", null, CommunityResponse::class.java)
    }

    /** Public reassurance stats. `country` (dial code) adds that country's counts. */
    fun stats(country: String? = null): StatsResponse {
        val countryParam = country?.takeIf { it.isNotBlank() }
            ?.let { "?country=${URLEncoder.encode(it, "UTF-8")}" } ?: ""
        val url = URL("$baseUrl/stats$countryParam")
        return request(url, "GET", null, StatsResponse::class.java)
    }

    /**
     * Short-lived access token gating GET /lists (the download URL is
     * deliberately "provisoire" — dead a few minutes later if replayed).
     */
    fun listsAccess(): ListAccessResponse {
        val url = URL("$baseUrl/lists/access")
        return request(url, "GET", null, ListAccessResponse::class.java)
    }

    /**
     * Delta sync. `since` = ISO-8601; null on first launch (warm-up = full list).
     * `country` = E.164 dial code (e.g. "33") to scope the list; null/blank = worldwide.
     * `token` = short-lived token from [listsAccess].
     */
    fun fetchList(since: String?, country: String? = null, limit: Int = 5000, token: String): ListResponse {
        val sinceParam = since?.let { "&since=${URLEncoder.encode(it, "UTF-8")}" } ?: ""
        val countryParam = country?.takeIf { it.isNotBlank() }
            ?.let { "&country=${URLEncoder.encode(it, "UTF-8")}" } ?: ""
        val tokenParam = "&token=${URLEncoder.encode(token, "UTF-8")}"
        val url = URL("$baseUrl/lists?limit=$limit$sinceParam$countryParam$tokenParam")
        return request(url, "GET", null, ListResponse::class.java)
    }

    /** Live lookup for a single number. */
    fun lookup(phone: String): LookupResponse {
        val url = URL("$baseUrl/lookup/${URLEncoder.encode(phone, "UTF-8")}")
        return request(url, "GET", null, LookupResponse::class.java)
    }

    /** "Numéros qui montent" — most-reported numbers ramping up recently. */
    fun trending(windowDays: Int = 7, limit: Int = 25): TrendingResponse {
        val url = URL("$baseUrl/trending?window=$windowDays&limit=$limit")
        return request(url, "GET", null, TrendingResponse::class.java)
    }

    /** Submit an anonymous report. */
    fun report(body: ReportRequest) {
        val url = URL("$baseUrl/reports")
        request(url, "POST", gson.toJson(body), Map::class.java)
    }

    /** Submit a daily-game score for `game` → returns today's leaderboard position. */
    fun submitGameScore(deviceId: String, score: Int, waves: Int, day: Long, game: String = "defense"): GameScoreResponse {
        val url = URL("$baseUrl/game/score")
        val body = GameScoreRequest(deviceId, score, waves, day, game)
        return request(url, "POST", gson.toJson(body), GameScoreResponse::class.java)
    }

    /** Anonymous leaderboard for a game + period (day|week|all) + your position. */
    fun gameLeaderboard(period: String, day: Long, deviceId: String, game: String = "defense"): LeaderboardResponse {
        val url = URL("$baseUrl/game/leaderboard?period=$period&day=$day&game=$game&deviceId=${URLEncoder.encode(deviceId, "UTF-8")}")
        return request(url, "GET", null, LeaderboardResponse::class.java)
    }

    /** This device's recent daily results for a game, with per-day rank. */
    fun gameHistory(deviceId: String, days: Int = 14, game: String = "defense"): HistoryResponse {
        val url = URL("$baseUrl/game/history?deviceId=${URLEncoder.encode(deviceId, "UTF-8")}&days=$days&game=$game")
        return request(url, "GET", null, HistoryResponse::class.java)
    }

    /** The day's TRACE sprint (5 grids), identical for every player. */
    fun tracePuzzle(day: Long): TracePuzzleSet {
        val url = URL("$baseUrl/game/puzzle?game=trace&day=$day")
        return request(url, "GET", null, TracePuzzleSet::class.java)
    }

    /** A single unranked TRACE training grid of the requested size. */
    fun traceTrainingPuzzle(size: Int, seed: Long): TracePuzzleSet {
        val url = URL("$baseUrl/game/puzzle?game=trace&mode=training&size=$size&seed=$seed")
        return request(url, "GET", null, TracePuzzleSet::class.java)
    }

    /** RGPD: erase all reports made from this device. */
    fun eraseDevice(deviceId: String) {
        val url = URL("$baseUrl/privacy/device/${URLEncoder.encode(deviceId, "UTF-8")}")
        request(url, "DELETE", null, Map::class.java)
    }

    /** RGPD: request erasure of a specific number from the community database. */
    fun eraseNumber(phone: String) {
        val url = URL("$baseUrl/privacy/number/${URLEncoder.encode(phone, "UTF-8")}")
        request(url, "DELETE", null, Map::class.java)
    }

    private fun <T> request(url: URL, method: String, body: String?, type: Class<T>): T {
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 15_000
            setRequestProperty("Accept", "application/json")
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }
        try {
            if (body != null) {
                conn.outputStream.use { it.write(body.toByteArray()) }
            }
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val text = stream?.bufferedReader()?.use { it.readText() } ?: ""
            if (code !in 200..299) {
                throw ApiException(code, text)
            }
            return gson.fromJson(text, type)
        } finally {
            conn.disconnect()
        }
    }
}

class ApiException(val code: Int, message: String) : Exception("HTTP $code: $message")
